### PR TITLE
Action archive

### PR DESCRIPTION
**Pull request by benhylau**
_WIP_

This adds tighter validations to input ZIPs and names everything to spec. Adds info logging.

### CODE REVIEW

File: integritybackend/asset_helper.py [hunk 1]
```diff
@@ -165,33 +161,36 @@ def get_internal_claim_fullpath(self, from_file):
...
             self._collection_prefix(collection_id),
-            f"action-{action}-output" if output else f"action-{action}",
+            f"action-{action_name}-output" if output else f"action-{action_name}",
         )
 
+    def get_action_dir(self, collection_id, action_name):
+        return os.path.join(self.internal_prefix, collection_id, f"action-{action_name}")
+
+    def get_tmp_action_dir(self, collection_id, action_name):
+        return os.path.join(self.dir_internal_tmp, collection_id, f"action-{action_name}")
```
Comment (benhylau) [inline integritybackend/asset_helper.py:189]: Mirroring paths for consistency

File: integritybackend/numbers.py [hunk 2]
```diff
@@ -1,68 +1,71 @@
...
+        file_util = FileUtil()
+        asset_cid = file_util.digest_cidv1(asset)
+        asset_tree_cid = file_util.digest_cidv1(asset_tree)
+        asset_tree_sha = file_util.digest_sha256(asset_tree)
+        author_cid = file_util.digest_cidv1(author)
 
-    r = requests.post(
-        _COMMIT_URL,
-        headers={"Authorization", f"Bearer {config.NUMBERS_API_KEY}"},
-        files=[
```
Comment (benhylau) [inline integritybackend/numbers.py:39]: @makeworld-the-better-one did these work for you before?

File: integritybackend/actions.py [hunk 3]
```diff
@@ -47,103 +48,116 @@ def archive(self, asset_fullpath: str, org_config: dict, collection_id: str):
...
         )
-
-        action_config = action.get('params')
-
-        if action_config["encryption"]["algo"] != "aes-256-cbc":
+        action_params = action.get('params')
+        if action_params["encryption"]["algo"] != "aes-256-cbc":
             raise Exception(
-                f"Encryption algo {action_config['encryption']['algo']} not implemented"
+                f"Encryption algo {action_params['encryption']['algo']} not implemented"
```
Comment (benhylau) [inline integritybackend/actions.py:67]: @YurkoWasHere let's keep things named according to what they are.

File: integritybackend/actions.py [hunk 4]
```diff
@@ -4,11 +4,12 @@
 from .file_util import FileUtil
 from .filecoin import Filecoin
 from .iscn import Iscn
+from .log_helper import LogHelper
```
Comment (benhylau) [inline integritybackend/actions.py:7]: From now on we use LogHelper. It ensures the correct configurations of the logging is applied.

File: integritybackend/actions.py [hunk 5]
```diff
@@ -47,103 +48,116 @@ def archive(self, asset_fullpath: str, org_config: dict, collection_id: str):
...
+        archive_dir = asset_helper.get_action_dir(collection_id, "archive")
         tmp_zip = shutil.copy2(zip_path, archive_dir)
 
-        # Verify zip contents are valid and expected
+        # Verify ZIP contents are valid and expected
         zip_listing = zip_util.listing(tmp_zip)
-        if len(zip_listing) > 3:
-            # Should only have three: content, recorder meta, content meta
+        if len(zip_listing) != 3:
+            # ZIP must contain three files: content, meta-content , meta-recorder
```
Comment (benhylau) [inline integritybackend/actions.py:84]: We wants exactly three files and added filename verify for all below.

File: integritybackend/actions.py [hunk 6]
```diff
@@ -47,103 +48,116 @@ def archive(self, asset_fullpath: str, org_config: dict, collection_id: str):
...
+        zip_dir = os.path.join(tmp_dir, content_sha_unverified)
         extracted_content = os.path.join(zip_dir, content_filename)
         file_util.create_dir(zip_dir)
         zip_util.extract_file(tmp_zip, content_filename, extracted_content)
 
-        # Generate content hashes
+        # Generate content hashes and verify
         content_sha = file_util.digest_sha256(extracted_content)
+        if content_sha != content_sha_unverified:
+            raise Exception(f"SHA-256 of content does not match file name: {zip_path}")
```
Comment (benhylau) [inline integritybackend/actions.py:116]: Here it checks that the sha actually matches.

File: integritybackend/actions.py [hunk 7]
```diff
@@ -47,103 +48,116 @@ def archive(self, asset_fullpath: str, org_config: dict, collection_id: str):
...
+            zip_util.append(
+                tmp_zip,
+                content_ots,
+                "proofs/" + os.path.basename(content_ots),
+            )
+            _logger.info(f"Content registered on opentimestamps: {content_ots}")
+        else:
+            _logger.info(f"Content registration on opentimestamps skipped")
+
+        # Get archive ZIP hashes
```
Comment (benhylau) [inline integritybackend/actions.py:134]: @makeworld-the-better-one the ZIP is named `sha(ZIP)` not `sha(asset)`.

**Review by YurkoWasHere — APPROVED** [review]
Great variable renames.

Like the use of LogHelper, should carry that through to other components too.

File: integritybackend/actions.py (hunk 7 above)
Comment (makeworld-the-better-one) [reply integritybackend/actions.py:134]: From what I [wrote down](https://github.com/starlinglab/integrity-backend/issues/67#issuecomment-1082406996) from our previous meeting I thought it was the other way, but either works.

**Review by makeworld-the-better-one — APPROVED** [review]
LGTM, thanks for improving all this code. I noticed tests aren't passing, although that was happening before this PR. Seems like a CI problem unrelated to any specific code. Not sure if you want to fix that here or elsewhere.

File: integritybackend/actions.py (hunk 7 above)
Comment (benhylau) [reply integritybackend/actions.py:134]: >Name of final ZIP (unencrypted and encrypted) is the hash of the asset / content file

Oh you're right. I may be wrong on this. Let me fix either the docs or the implementation here and get back to you.

File: integritybackend/log_helper.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import logging
+import sys
+
+
+class LogHelper:
+
+    @staticmethod
+    def getLogger():
+        logging.basicConfig(
```
[the remark this answers is not in the document]
Comment (benhylau) [reply integritybackend/log_helper.py:9]: The problem is, it didn't work. This happened several times now and I think is bc of the subprocesses. I fixed it before by calling the init code in all the right places, but then something changed and logs disappeared again. The implementation now ensures it won't happen again as we'd centralize the config code, but apply it redundantly whenever an instance is fetched.

[the remark this answers is not in the document]
Comment (makeworld-the-better-one) [reply integritybackend/log_helper.py:9]: Hmm I guess that makes sense if by default processes aren't copying the memory or something. Anyway I understand now, this definitely seems more resilient then.

File: integritybackend/actions.py (hunk 7 above)
Comment (benhylau) [reply integritybackend/actions.py:134]: @YurkoWasHere brought up a good point that we'd have collision if two same assets have different metadata in a zip somehow. I think this is an important consideration.

**Pull request merged**